firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -69,32 +67,44 @@ void run() {
       }
     }
 
-    void add(Runnable callback) {
+    synchronized void add(Runnable callback) {
       callbacks.add(callback);
     }
 
     private final List<Runnable> callbacks = new ArrayList<>();
   }
 
   public static class StopListenerSupportFragment extends android.support.v4.app.Fragment {
+    CallbackList callbacks = new CallbackList();
+
     @Override
     public void onStop() {
       super.onStop();
-      callbacks.run();
-    }
 
-    CallbackList callbacks = new CallbackList();
+      CallbackList callbacksCopy;
+      synchronized (callbacks) {
```

Comment:
Based on in-person discussion this does make sense since `callbacks.add` is itself now synchronized.

I think it's worth adding a comment here about what we're defending against so that's more obvious why the synchronization isn't on the `StopListenerSupportFragment` instance and it's OK for successive invocations of `onStop` to synchronize on different objects (which otherwise seems like a bug).

Reply:
Minor nit: since all accesses/modifications to the callbacks now happens on the UI thread there is no need to synchronize at all, e.g.:
* Callback registration runs on UI thread
* Fragment onStop runs on UI thread.

It is fine to leave as is though.

Reply:
Yeah, I noticed that too, but decided to leave it in because the fact that the registration happens on the ui thread is an implementation detail; in a few of the prior implementations we were discussing in this PR, it didn't happen in the UI thread at all. If we go back to one of those other implementations, we might miss the necessity to sync here. (Indeed, I missed that in my first pass.)